### DISCUSSION TITLE
How to configure number of epochs and threshold in PatchCore

### DISCUSSION THREAD

**Discussion by enislalmi** [open-edge-platform/anomalib#1252] [Q&A]
### Describe the bug

I have the config file as follows and I am using a custom dataset. The first Epoch runs fine, and then it crashes giving this error. As you can see I have set treshold: ... adaptive=false, as I saw in another issue, but it still is not working for me. Any ideas? Thanks!

### Dataset

Other (please specify in the text field below)

### Model

PatchCore

### Steps to reproduce the behavior

1. Create the config file as above.

2. Read the data as cited in the notebooks:
`folder_dataset_classification_train = FolderDataset(
    normal_dir=folder_dataset_root / "train",
    abnormal_dir=folder_dataset_root / "test/reject",
    split="train",
    transform=transform,
    task="classification",
)
folder_dataset_classification_train.setup()
folder_dataset_classification_train.samples.head()`  (similar for test)

3. Get the config file from the specified path:
`config = get_configurable_parameters(
    model_name="patchcore", config_path=path1
)
if config.project.seed:
    seed_everything(config.project.seed)
datamodule = get_datamodule(config)
model1 = get_model(config)

trainer = Trainer(**config.trainer)

trainer.fit(model=model1, datamodule=datamodule)`

### OS information

OS information:
- OS: [Microsoft 11]
- Python version: [3.10]
- Anomalib version: [latest pull]
- PyTorch version: ['2.0.0+cu117'
- GPU models and configuration: [e.g. NVIDIA GeForce RTX 3050]
- Any other relevant information: [e.g. I'm using a custom dataset]

### Expected behavior

The trainer should run until the end, not just for epoch 0.

### Pip/GitHub

pip

### What version/branch did you use?

latest

### Configuration YAML

```yaml
dataset:
  name: data1
  format: folder
  path: ./datasets/data1
  normal_dir: train/good
  abnormal_dir: test/reject
  normal_test_dir: test/good
  task: classification
  mask: null
  extensions: null
  normalization: imagenet # data distribution to which the images will be normalized: [none, imagenet]
  split_ratio: 0.2 # ratio of the normal images that will be used to create a test split
  image_size: 64
  train_batch_size: 32
  test_batch_size: 32
  num_workers: 8
  transform_config:
    train: null
    val: null
  test_split_mode: from_dir # options: [from_dir, synthetic]
  test_split_ratio: 0.2 # fraction of train images held out testing (usage depends on test_split_mode)
  val_split_mode: same_as_test # options: [same_as_test, from_test, synthetic]
  val_split_ratio: 0.5 # fraction of train/test images held out for validation (usage depends on val_split_mode)
  create_validation_set: true
  tiling:
    apply: false
    tile_size: null
    stride: null
    remove_border_count: 0
    use_random_tiling: False
    random_tile_count: 16

model:
  name: patchcore
  backbone: wide_resnet50_2
  pre_trained: true
  layers:
    - layer2
    - layer3
  coreset_sampling_ratio: 0.1
  num_neighbors: 9
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null
    adaptive: false

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]

project:
  seed: 0
  path: models

logging:
  logger: [] # options: [comet, tensorboard, wandb, csv] or combinations.
  log_graph: false # Logs the model graph to respective logger.

optimization:
  export_mode: null # options: onnx, openvino

# PL Trainer Args. Don't add extra parameter here.
trainer:
  accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
  accumulate_grad_batches: 1
  amp_backend: native
  auto_lr_find: false
  auto_scale_batch_size: false
  auto_select_gpus: false
  benchmark: false
  check_val_every_n_epoch: 1 # Don't validate before extracting features.
  default_root_dir: null
  detect_anomaly: false
  deterministic: false
  devices: 1
  enable_checkpointing: true
  enable_model_summary: true
  enable_progress_bar: true
  fast_dev_run: false
  gpus: null # Set automatically
  gradient_clip_val: 0
  ipus: null
  limit_predict_batches: 1.0
  limit_test_batches: 1.0
  limit_train_batches: 1.0
  limit_val_batches: 1.0
  log_every_n_steps: 50
  max_epochs: 1
  max_steps: -1
  max_time: null
  min_epochs: null
  min_steps: null
  move_metrics_to_cpu: false
  multiple_trainloader_mode: max_size_cycle
  num_nodes: 1
  num_processes: null
  num_sanity_val_steps: 0
  overfit_batches: 0.0
  plugins: null
  precision: 32
  profiler: null
  reload_dataloaders_every_n_epochs: 0
  replace_sampler_ddp: true
  strategy: null
  sync_batchnorm: false
  tpu_cores: null
  track_grad_norm: -1
  val_check_interval: 1.0 # Don't validate before extracting features.
```

### Logs

```shell
n/a
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by samet-akcay** [comment]
@enislalmi, if I understand your question, there are two aspects of this problem. 
1. Number of Epochs
For `patchcore` you should set the number of epochs to 1. This is mainly because patch core does not train a CNN model. Instead, it extracts the features from the CNN. Increasing the number of epochs does not have any impact.

1. Adaptive Threshold
If you would like to change the threshold method you should configure the `method` parameter:

```yaml
  threshold:
    method: adaptive #options: [adaptive, manual]
    manual_image: null
    manual_pixel: null
```

Note that, if you set `method` to `manual`, you also need to provide manual image/pixel threshold values.
